misc/python/materialize/cli/orchestratord.py

```diff
@@ -215,74 +216,124 @@ def root_psql(cmd: str):
-                resource_id = None
-        except subprocess.CalledProcessError:
-            resource_id = None
+    resource_id = get_resource_id(args)
+    retry(
+        lambda: env_kubectl(
+            "wait",
+            "--for=condition=Available",
+            f"deployment/mz{resource_id}-console",
+        ),
+        exception_types=(subprocess.CalledProcessError,),
+    )
 
-        if resource_id is not None:
-            break
 
-        sleep(1)
-    assert resource_id is not None
+def portforward(args: argparse.Namespace):
+    env_kubectl = make_env_kubectl(args)
 
-    node_port = None
-    for _ in range(60):
-        try:
-            node_port = (
-                env_kubectl(
-                    "get",
-                    "service",
-                    f"mz{resource_id}-console",
-                    '-o=jsonpath={.spec.ports[?(@.name=="http")].nodePort}',
-                )
-                .decode()
-                .strip()
+    resource_id = get_resource_id(args)
+
+    sockets = []
+    try:
+        for _ in range(5):
+            s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+            sockets.append(s)
+            s.bind(("", 0))
+            s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+        ports = [s.getsockname()[1] for s in sockets]
+    finally:
+        for s in sockets:
+            s.close()
+
+    port_destinations = {}
+
+    def port_forward(target: str, target_port: str):
+        port = ports.pop()
+        port_destinations[f"{target}/{target_port}"] = port
+
+        # in theory doing a port forward to the service directly should work,
+        # but i can't seem to get it to
+        selector = json.loads(
+            env_kubectl(
+                "get",
+                "service",
+                f"mz{resource_id}-{target}",
+                "-o=jsonpath={.spec.selector}",
             )
-            if node_port == "<no value>" or node_port == "":
-                node_port = None
-        except subprocess.CalledProcessError:
-            node_port = None
+        )
+        pod = env_kubectl(
+            "get",
+            "pod",
+            "-o=name",
+            *[f"--selector={k}={v}" for k, v in selector.items()],
+        ).splitlines()[0]
 
-        if node_port is not None:
-            break
+        env_kubectl(
+            "port-forward",
+            pod,
+            f"{port}:{target_port}",
+        )
+
+    threads = [
+        threading.Thread(target=port_forward, args=["console", "http"]),
+        threading.Thread(target=port_forward, args=["balancerd", "pgwire"]),
+        threading.Thread(target=port_forward, args=["balancerd", "http"]),
+        threading.Thread(target=port_forward, args=["environmentd", "internal-sql"]),
+        threading.Thread(target=port_forward, args=["environmentd", "internal-http"]),
+    ]
+
+    for t in threads:
+        t.start()
+    while len(port_destinations) < len(threads):
+        pass
```

Comment:
This isn't actually waiting for the port forward to be complete, only that we've started each of the threads. We don't actually join the threads. That may introduce surprising behavior, as this function may return early, and downstream things trying to use the port forward may fail.

Reply:
yeah, the way port forwarding works is that the `kubectl port-forward` command doesn't return, and the port forward is only active while that command is holding a connection open, and as far as i can tell, there is no way to actually query kubernetes to tell whether a port forward is active. i just remove the auto-open behavior, which is probably as close as we're going to get.